book/src/dev/rfcs/0010-v5-transaction.md

```diff
@@ -146,20 +146,49 @@ trait AnchorVariant {
 [changes-to-sapling-shieldeddata]: #changes-to-sapling-shieldeddata
 
 We use `AnchorVariant` in `ShieldedData` to model the anchor differences between V4 and V5:
+* in V4, there is a per-spend anchor
+* in V5, there is a shared anchor, which is only present when there are spends
+
+If there are no spends and no outputs:
+* in v4, the value_balance is fixed to zero
+* in v5, the value balance field is not present
+* in both versions, the binding_sig field is not present
 
 ```rust
+/// ShieldedData ensures that value_balance and binding_sig are only present when
+/// there is at least one spend or output.
 struct sapling::ShieldedData<AnchorV: AnchorVariant> {
     value_balance: Amount,
-    shared_anchor: AnchorV::Shared,
-    // The following fields are in a different order to the serialized data, see:
-    // https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus
-    first: Either<Spend<AnchorV>, Output>,
-    rest_spends: Vec<Spend<AnchorV>>,
-    rest_outputs: Vec<Output>,
+    transfers: sapling::TransferData<AnchorV>,
     binding_sig: redjubjub::Signature<Binding>,
 }
+
+/// TransferData ensures that:
+/// * there is at least one spend or output, and
+/// * the shared anchor is only present when there are spends
+enum sapling::TransferData<AnchorV: AnchorVariant> {
+    /// In Transaction::V5, if there are any spends,
+    /// there must also be a shared spend anchor.
+    Spends {
```

Comment:
Thinking about this a bit more, we might want to call these variants:
* SpendsAndMaybeOutputs
* JustOutputs

We only reference them a few times, and only inside the sapling module, so the long names shouldn't be an issue for any new development work.